File: ports/zig/src/widgets/text.zig
//! Text, badges, label/value pairs, dividers and the function-key bar.

const std = @import("std");

const buffer_mod = @import("../buffer.zig");
const color_mod = @import("../color.zig");
const surface_mod = @import("../surface.zig");
const theme_mod = @import("../theme.zig");
const unicode = @import("../unicode.zig");

const Align = unicode.Align;
const Attrs = buffer_mod.Attrs;
const Color = color_mod.Color;
const Style = buffer_mod.Style;
const Surface = surface_mod.Surface;
const TextOptions = surface_mod.TextOptions;

pub const TextStyle = struct {
    fg: ?Color = null,
    bg: ?Color = null,
    attrs: ?Attrs = null,
    alignment: Align = .left,
    wrap: bool = false,
    bold: bool = false,
    dim: bool = false,
    italic: bool = false,
    underline: bool = false,

    fn resolvedAttrs(self: TextStyle) Attrs {
        var a = self.attrs orelse Attrs.none;
        if (self.bold) a.bold = true;
        if (self.dim) a.dim = true;
        if (self.italic) a.italic = true;
        if (self.underline) a.underline = true;
        return a;
    }
};

/// `allocator` is only touched when `style.wrap` is set; unwrapped text splits
/// on newlines in place.
pub fn drawText(
    allocator: std.mem.Allocator,
    s: Surface,
    content: []const u8,
    style: TextStyle,
) !void {
    if (s.isEmpty()) return;
    const options: TextOptions = .{
        .fg = style.fg orelse s.theme.foreground,
        .bg = style.bg,
        .attrs = style.resolvedAttrs(),
    };

    var scratch: [4096]u8 = undefined;
    var padded: [4096]u8 = undefined;

    if (style.wrap) {
        const lines = try unicode.wrap(allocator, content, s.width());
        defer unicode.freeWrapped(allocator, lines);
        for (lines, 0..) |line, i| {
            if (i >= s.height()) break;
            const cut = unicode.truncate(&scratch, line, s.width());
            const shown = unicode.fit(&padded, cut, s.width(), style.alignment);
            _ = s.text(0, @intCast(i), shown, options);
        }
        return;
    }

    var it = std.mem.splitScalar(u8, content, '\n');
    var i: usize = 0;
    while (it.next()) |line| : (i += 1) {
        if (i >= s.height()) break;
        const cut = unicode.truncate(&scratch, line, s.width());
        const shown = unicode.fit(&padded, cut, s.width(), style.alignment);
        _ = s.text(0, @intCast(i), shown, options);
    }
}

/// Filled reads as a chip; outline keeps the panel quiet.
pub const BadgeVariant = enum { filled, outline, subtle };

pub const BadgeOptions = struct {
    text: []const u8 = "",
    color: ?Color = null,
    variant: BadgeVariant = .filled,
    alignment: Align = .left,
};

/// Returns the width drawn, so callers can advance a cursor past it.
pub fn drawBadge(s: Surface, options: BadgeOptions) usize {
    if (s.isEmpty()) return 0;
    const theme = s.theme;
    const c = options.color orelse theme.primary;

    var label_buf: [256]u8 = undefined;
    const label = std.fmt.bufPrint(&label_buf, " {s} ", .{options.text}) catch return 0;

    const style: Style = switch (options.variant) {
        .filled => .{
            .fg = if (theme.dark) theme.background else theme.surface,
            .bg = c,
            .attrs = .bold_only,
        },
        .subtle => .{ .fg = c, .bg = theme.surface.mix(c, 0.18) },
        .outline => .{ .fg = c, .attrs = .bold_only },
    };

    const width = @min(unicode.stringWidth(label), s.width());
    const x: isize = switch (options.alignment) {
        .right => @as(isize, @intCast(s.width())) - @as(isize, @intCast(width)),
        .center => @divFloor(@as(isize, @intCast(s.width())) - @as(isize, @intCast(width)), 2),
        .left => 0,
    };
    var cut: [256]u8 = undefined;
    _ = s.text(@max(0, x), 0, unicode.truncate(&cut, label, s.width()), TextOptions.fromStyle(style));
    return width;
}

pub const KeyValueRow = struct {
    label: []const u8,
    value: []const u8,
    color: ?Color = null,
    label_color: ?Color = null,
};

pub const KeyValueOptions = struct {
    rows: []const KeyValueRow = &.{},
    /// Columns reserved for labels. Null means the widest label.
    label_width: ?usize = null,
    gap: usize = 1,
    /// Push values to the right edge instead of next to the label.
    spread: bool = true,
    label_color: ?Color = null,
    value_color: ?Color = null,
    background: ?Color = null,
};

/// Aligned label/value pairs — the backbone of every "System" panel.
pub fn drawKeyValues(s: Surface, options: KeyValueOptions) void {
    if (s.isEmpty()) return;
    const theme = s.theme;

    const label_width = options.label_width orelse blk: {
        var widest: usize = 0;
        for (options.rows) |r| widest = @max(widest, unicode.stringWidth(r.label));
        const cap = @max(4, @as(usize, @intFromFloat(@as(f64, @floatFromInt(s.width())) * 0.6)));
        break :blk @min(widest + 1, cap);
    };

    for (options.rows, 0..) |row, i| {
        if (i >= s.height()) break;
        var cut: [256]u8 = undefined;
        var padded: [256]u8 = undefined;
        const label = unicode.fit(
            &padded,
            unicode.truncate(&cut, row.label, label_width),
            label_width,
            .left,
        );
        _ = s.text(0, @intCast(i), label, .{
            .fg = row.label_color orelse options.label_color orelse theme.muted,
            .bg = options.background,
        });

        const vx = label_width + options.gap;
        const vw = s.width() -| vx;
        if (vw == 0) continue;

        var value_cut: [256]u8 = undefined;
        var value_padded: [256]u8 = undefined;
        var value = unicode.truncate(&value_cut, row.value, vw);
        if (options.spread) value = unicode.fit(&value_padded, value, vw, .right);
        _ = s.text(@intCast(vx), @intCast(i), value, .{
            .fg = row.color orelse options.value_color orelse theme.foreground,
            .bg = options.background,
        });
    }
}

pub const DividerOptions = struct {
    label: []const u8 = "",
    color: ?Color = null,
    char: u21 = '─',
    alignment: Align = .left,
};

pub fn drawDivider(s: Surface, options: DividerOptions) void {
    if (s.isEmpty()) return;
    const theme = s.theme;
    s.hline(0, 0, s.width(), options.char, .{ .fg = options.color orelse theme.border });
    if (options.label.len == 0) return;

    var label_buf: [256]u8 = undefined;
    const label = std.fmt.bufPrint(&label_buf, " {s} ", .{options.label}) catch return;
    const w: isize = @intCast(unicode.stringWidth(label));
    const width: isize = @intCast(s.width());
    const x: isize = switch (options.alignment) {
        .left => 1,
        .right => width - w - 1,
        .center => @divFloor(width - w, 2),
    };
    var cut: [256]u8 = undefined;
    _ = s.text(@max(0, x), 0, unicode.truncate(&cut, label, s.width()), .{ .fg = theme.muted });
}

pub const StatusItem = struct {
    label: []const u8 = "",
    key: []const u8 = "",
    color: ?Color = null,
    /// Highlight this entry, e.g. the active tab or a live indicator.
    active: bool = false,
};

/// Caps reverse-videos the key caps, like a function-key bar.
pub const KeyStyle = enum { caps, plain };

pub const StatusBarOptions = struct {
    items: []const StatusItem = &.{},
    right: []const StatusItem = &.{},
    background: ?Color = null,
    key_color: ?Color = null,
    key_style: KeyStyle = .caps,
};

/// The F1/F2/F10 bar along the bottom of every serious TUI.
pub fn drawStatusBar(s: Surface, options: StatusBarOptions) void {
    if (s.isEmpty()) return;
    const theme = s.theme;
    const bg = options.background orelse theme_mod.elevate(theme.*, 0.04);
    s.fill(.{ .bg = bg });
    const key_color = options.key_color orelse theme.accent;

    const drawItems = struct {
        fn f(
            surface: Surface,
            items: []const StatusItem,
            start_x: isize,
            bar_bg: Color,
            cap_color: Color,
            key_style: KeyStyle,
        ) isize {
            const t = surface.theme;
            var cx = start_x;
            for (items) |item| {
                if (cx >= @as(isize, @intCast(surface.width()))) break;
                if (item.key.len > 0) {
                    const cap: Style = switch (key_style) {
                        .plain => .{ .fg = cap_color, .bg = bar_bg, .attrs = .bold_only },
                        .caps => .{
                            .fg = if (t.dark) t.background else t.surface,
                            .bg = cap_color,
                            .attrs = .bold_only,
                        },
                    };
                    cx += @intCast(surface.text(cx, 0, item.key, TextOptions.fromStyle(cap)));
                    cx += @intCast(surface.text(cx, 0, " ", .{ .bg = bar_bg }));
                }
                cx += @intCast(surface.text(cx, 0, item.label, .{
                    .fg = if (item.active) t.foreground else (item.color orelse t.muted),
                    .bg = bar_bg,
                    .attrs = if (item.active) Attrs.bold_only else Attrs.none,
                }));
                cx += @intCast(surface.text(cx, 0, "  ", .{ .bg = bar_bg }));
            }
            return cx;
        }
    }.f;

    const x = drawItems(s, options.items, 1, bg, key_color, options.key_style);

    if (options.right.len > 0) {
        var width: usize = 0;
        for (options.right) |i| {
            width += unicode.stringWidth(i.label) + 2;
            if (i.key.len > 0) width += unicode.stringWidth(i.key) + 1;
        }
        const from = @max(x, @as(isize, @intCast(s.width())) - @as(isize, @intCast(width)) - 1);
        _ = drawItems(s, options.right, from, bg, key_color, options.key_style);
    }
}

File: ports/rust/src/widgets/text.rs
//! Text, badges, label/value pairs, dividers and the function-key bar.

use crate::buffer::{Attrs, Style};
use crate::color::Color;
use crate::surface::{Surface, TextOptions};
use crate::theme::elevate;
use crate::unicode::{drop_columns, fit, string_width, truncate, wrap, Align};

#[derive(Clone, Debug, Default)]
pub struct TextStyle {
    pub fg: Option<Color>,
    pub bg: Option<Color>,
    pub attrs: Option<Attrs>,
    pub align: Option<Align>,
    pub wrap: bool,
    pub bold: bool,
    pub dim: bool,
    pub italic: bool,
    pub underline: bool,
    /// First line to show, counted after wrapping.
    ///
    /// After wrapping is the only place this can be correct: the caller does
    /// not know how many lines their text became, and pre-slicing the string
    /// means re-deciding every time the width changes.
    pub scroll: usize,
    /// Columns to shift the text left by, for lines wider than the surface.
    pub scroll_x: usize,
}

impl TextStyle {
    pub fn new() -> TextStyle {
        TextStyle::default()
    }

    pub fn fg(mut self, c: Color) -> TextStyle {
        self.fg = Some(c);
        self
    }

    pub fn bg(mut self, c: Color) -> TextStyle {
        self.bg = Some(c);
        self
    }

    pub fn align(mut self, a: Align) -> TextStyle {
        self.align = Some(a);
        self
    }

    pub fn wrapped(mut self) -> TextStyle {
        self.wrap = true;
        self
    }

    pub fn bold(mut self) -> TextStyle {
        self.bold = true;
        self
    }

    pub fn dim(mut self) -> TextStyle {
        self.dim = true;
        self
    }

    pub fn italic(mut self) -> TextStyle {
        self.italic = true;
        self
    }

    pub fn underline(mut self) -> TextStyle {
        self.underline = true;
        self
    }

    fn resolved_attrs(&self) -> Attrs {
        let mut a = self.attrs.unwrap_or(Attrs::NONE);
        if self.bold {
            a |= Attrs::BOLD;
        }
        if self.dim {
            a |= Attrs::DIM;
        }
        if self.italic {
            a |= Attrs::ITALIC;
        }
        if self.underline {
            a |= Attrs::UNDERLINE;
        }
        a
    }
}

pub fn draw_text(surface: &Surface, content: &str, options: &TextStyle) {
    if surface.is_empty() {
        return;
    }
    let style = Style {
        fg: Some(options.fg.unwrap_or(surface.theme.foreground)),
        bg: options.bg,
        attrs: Some(options.resolved_attrs()),
    };
    let lines: Vec<String> = if options.wrap {
        wrap(content, surface.width())
    } else {
        content.split('\n').map(String::from).collect()
    };
    let align = options.align.unwrap_or(Align::Left);
    let visible = if options.scroll < lines.len() { &lines[options.scroll..] } else { &[][..] };
    for (i, line) in visible.iter().enumerate().take(surface.height()) {
        let shifted;
        let line = if options.scroll_x > 0 {
            shifted = drop_columns(line, options.scroll_x);
            &shifted
        } else {
            line
        };
        let padded = fit(&truncate(line, surface.width()), surface.width(), align);
        surface.text(0, i as isize, &padded, &TextOptions::from(style));
    }
}

/// Filled reads as a chip; outline keeps the panel quiet.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub enum BadgeVariant {
    #[default]
    Filled,
    Outline,
    Subtle,
}

#[derive(Clone, Debug, Default)]
pub struct BadgeOptions {
    pub text: String,
    pub color: Option<Color>,
    pub variant: BadgeVariant,
    pub align: Option<Align>,
}

impl BadgeOptions {
    pub fn new(text: impl Into<String>) -> BadgeOptions {
        BadgeOptions { text: text.into(), ..Default::default() }
    }

    pub fn color(mut self, c: Color) -> BadgeOptions {
        self.color = Some(c);
        self
    }

    pub fn variant(mut self, v: BadgeVariant) -> BadgeOptions {
        self.variant = v;
        self
    }
}

/// Returns the width drawn, so callers can advance a cursor past it.
pub fn draw_badge(surface: &Surface, options: &BadgeOptions) -> usize {
    if surface.is_empty() {
        return 0;
    }
    let theme = surface.theme.clone();
    let color = options.color.unwrap_or(theme.primary);
    let label = format!(" {} ", options.text);
    let style = match options.variant {
        BadgeVariant::Filled => Style {
            fg: Some(if theme.dark { theme.background } else { theme.surface }),
            bg: Some(color),
            attrs: Some(Attrs::BOLD),
        },
        BadgeVariant::Subtle => Style {
            fg: Some(color),
            bg: Some(theme.surface.mix(color, 0.18)),
            attrs: None,
        },
        BadgeVariant::Outline => Style { fg: Some(color), bg: None, attrs: Some(Attrs::BOLD) },
    };
    let width = string_width(&label).min(surface.width());
    let x = match options.align.unwrap_or(Align::Left) {
        Align::Right => surface.width() as isize - width as isize,
        Align::Center => (surface.width() as isize - width as isize) / 2,
        Align::Left => 0,
    };
    surface.text(x.max(0), 0, &truncate(&label, surface.width()), &TextOptions::from(style));
    width
}

#[derive(Clone, Debug, Default)]
pub struct KeyValueRow {
    pub label: String,
    pub value: String,
    pub color: Option<Color>,
    pub label_color: Option<Color>,
}

impl KeyValueRow {
    pub fn new(label: impl Into<String>, value: impl Into<String>) -> KeyValueRow {
        KeyValueRow { label: label.into(), value: value.into(), ..Default::default() }
    }

    pub fn color(mut self, c: Color) -> KeyValueRow {
        self.color = Some(c);
        self
    }
}

#[derive(Clone, Debug, Default)]
pub struct KeyValueOptions {
    pub rows: Vec<KeyValueRow>,
    /// Columns reserved for labels. Defaults to the widest label.
    pub label_width: Option<usize>,
    pub gap: Option<usize>,
    /// Push values to the right edge instead of next to the label. Default on.
    pub spread: Option<bool>,
    pub label_color: Option<Color>,
    pub value_color: Option<Color>,
    pub background: Option<Color>,
}

/// Aligned label/value pairs — the backbone of every "System" panel.
pub fn draw_key_values(surface: &Surface, options: &KeyValueOptions) {
    if surface.is_empty() {
        return;
    }
    let theme = surface.theme.clone();
    let gap = options.gap.unwrap_or(1);
    let label_width = options.label_width.unwrap_or_else(|| {
        let widest = options.rows.iter().map(|r| string_width(&r.label)).max().unwrap_or(0);
        (widest + 1).min((4).max((surface.width() as f64 * 0.6).floor() as usize))
    });

    for (i, row) in options.rows.iter().enumerate().take(surface.height()) {
        surface.text(
            0,
            i as isize,
            &fit(&truncate(&row.label, label_width), label_width, Align::Left),
            &TextOptions {
                fg: Some(row.label_color.or(options.label_color).unwrap_or(theme.muted)),
                bg: options.background,
                ..Default::default()
            },
        );
        let vx = label_width + gap;
        let vw = surface.width().saturating_sub(vx);
        if vw == 0 {
            continue;
        }
        let value = truncate(&row.value, vw);
        let shown = if options.spread == Some(false) {
            value
        } else {
            fit(&value, vw, Align::Right)
        };
        surface.text(
            vx as isize,
            i as isize,
            &shown,
            &TextOptions {
                fg: Some(row.color.or(options.value_color).unwrap_or(theme.foreground)),
                bg: options.background,
                ..Default::default()
            },
        );
    }
}

#[derive(Clone, Debug, Default)]
pub struct DividerOptions {
    pub label: Option<String>,
    pub color: Option<Color>,
    pub char: Option<char>,
    pub align: Option<Align>,
}

pub fn draw_divider(surface: &Surface, options: &DividerOptions) {
    if surface.is_empty() {
        return;
    }
    let theme = surface.theme.clone();
    let color = options.color.unwrap_or(theme.border);
    let ch = options.char.unwrap_or('─');
    surface.hline(0, 0, surface.width(), ch, &Style::new().with_fg(color));
    if let Some(label) = &options.label {
        let text = format!(" {label} ");
        let w = string_width(&text);
        let x = match options.align.unwrap_or(Align::Left) {
            Align::Left => 1isize,
            Align::Right => surface.width() as isize - w as isize - 1,
            Align::Center => (surface.width() as isize - w as isize) / 2,
        };
        surface.text(
            x.max(0),
            0,
            &truncate(&text, surface.width()),
            &TextOptions::new().fg(theme.muted),
        );
    }
}

#[derive(Clone, Debug, Default)]
pub struct StatusItem {
    pub key: Option<String>,
    pub label: String,
    pub color: Option<Color>,
    /// Highlight this entry, e.g. the active tab or a live indicator.
    pub active: bool,
}

impl StatusItem {
    pub fn new(label: impl Into<String>) -> StatusItem {
        StatusItem { label: label.into(), ..Default::default() }
    }

    pub fn key(mut self, key: impl Into<String>) -> StatusItem {
        self.key = Some(key.into());
        self
    }

    pub fn active(mut self) -> StatusItem {
        self.active = true;
        self
    }
}

/// Reverse-video the key caps, like a function-key bar, or leave them plain.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub enum KeyStyle {
    #[default]
    Caps,
    Plain,
}

#[derive(Clone, Debug, Default)]
pub struct StatusBarOptions {
    pub items: Vec<StatusItem>,
    pub right: Vec<StatusItem>,
    pub background: Option<Color>,
    pub key_color: Option<Color>,
    pub key_style: KeyStyle,
}

/// The F1/F2/F10 bar along the bottom of every serious TUI.
pub fn draw_status_bar(surface: &Surface, options: &StatusBarOptions) {
    if surface.is_empty() {
        return;
    }
    let theme = surface.theme.clone();
    let bg = options.background.unwrap_or_else(|| elevate(&theme, 0.04));
    surface.fill(&Style::new().with_bg(bg));

    let draw_items = |items: &[StatusItem], start_x: isize| -> isize {
        let mut cx = start_x;
        for item in items {
            if cx >= surface.width() as isize {
                break;
            }
            if let Some(key) = &item.key {
                let cap_style = match options.key_style {
                    KeyStyle::Plain => Style {
                        fg: Some(options.key_color.unwrap_or(theme.accent)),
                        bg: Some(bg),
                        attrs: Some(Attrs::BOLD),
                    },
                    KeyStyle::Caps => Style {
                        fg: Some(if theme.dark { theme.background } else { theme.surface }),
                        bg: Some(options.key_color.unwrap_or(theme.accent)),
                        attrs: Some(Attrs::BOLD),
                    },
                };
                cx += surface.text(cx, 0, key, &TextOptions::from(cap_style)) as isize;
                cx += surface.text(cx, 0, " ", &TextOptions::new().bg(bg)) as isize;
            }
            cx += surface.text(
                cx,
                0,
                &item.label,
                &TextOptions {
                    fg: Some(if item.active {
                        theme.foreground
                    } else {
                        item.color.unwrap_or(theme.muted)
                    }),
                    bg: Some(bg),
                    attrs: Some(if item.active { Attrs::BOLD } else { Attrs::NONE }),
                    ..Default::default()
                },
            ) as isize;
            cx += surface.text(cx, 0, "  ", &TextOptions::new().bg(bg)) as isize;
        }
        cx
    };

    let x = draw_items(&options.items, 1);

    if !options.right.is_empty() {
        let width: usize = options
            .right
            .iter()
            .map(|i| {
                string_width(&i.label)
                    + i.key.as_ref().map(|k| string_width(k) + 1).unwrap_or(0)
                    + 2
            })
            .sum();
        let from = x.max(surface.width() as isize - width as isize - 1);
        draw_items(&options.right, from);
    }
}
